### PR TITLE
451006-YauheniiKhudzik

### PR DESCRIPTION
**Pull request by MartinEmbers**
Please, take a look

### CODE REVIEW

File: Mapper.Tests/CacheTests/CachedMapperCollectionTests/CachedMapperCollectionTests.cs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        {
+            IEqualityComparer<IMappingUnit> comparer = new FakeMapperEqualityComparer();
+
+            ICachedMapperCollection collection = CreateCachedMapperCollection(comparer);
+
+            Assert.False(collection.ContainsKey(CreateMappingUnit()));
+        }
+
+        [Test]
+        public void ContainsKey_MissingElementPassed_FalseReturned()
```
Comment (RamanBut-Husaim) [inline Mapper.Tests/CacheTests/CachedMapperCollectionTests/CachedMapperCollectionTests.cs]: @MartinEmbers You could create a bunch of unit-tests there with the help of `TestCase`.

File: Mapper.Tests/CacheTests/CachedMapperCollectionTests/FakeMapperEqualityComparer.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+﻿using System.Collections.Generic;
+using Mapper.UnitsForMapping;
+
+namespace Mapper.Tests.CacheTests.CachedMapperCollectionTests
+{
+    internal class FakeMapperEqualityComparer : IEqualityComparer<IMappingUnit>
```
Comment (RamanBut-Husaim) [inline Mapper.Tests/CacheTests/CachedMapperCollectionTests/FakeMapperEqualityComparer.cs]: @MartinEmbers You could use [moq](https://github.com/Moq/moq4/wiki/Quickstart) or other library to create fake objects. Do not change current implementation but be aware of what could be done.

File: Mapper.Tests/SimpleMapperTests/SimpleMapperTests.cs [hunk 3]
```diff
@@ -0,0 +1,106 @@
+﻿using System;
+using Mapper.Cache;
+using Mapper.Compilers;
+using Mapper.Configuration;
+using Mapper.UnitsForMapping;
+using NSubstitute;
```
Comment (RamanBut-Husaim) [inline Mapper.Tests/SimpleMapperTests/SimpleMapperTests.cs]: @MartinEmbers Ok. `NSubstitute` is used.

File: Mapper.Tests/UtilsTests/TypeUtilsTests/TypeUtilsTests.cs [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+            Assert.Catch(() => { TypeUtils.IsConvertibleTypes(null, typeof(int)); });
+        }
+
+        [Test]
+        public void IsConvertibleTypes_RightParamNullPassed_ExceptionThrown()
+        {
+            Assert.Catch(() => { TypeUtils.IsConvertibleTypes(typeof(int), null); });
+        }
+
+        /*                        case TypeCode.UInt16:
```
Comment (RamanBut-Husaim) [inline Mapper.Tests/UtilsTests/TypeUtilsTests/TypeUtilsTests.cs]: @MartinEmbers Remove dead code please.

File: Mapper/SimpleMapper.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    public class SimpleMapper : IMapper
+    {
+        private readonly ICachedMapperCollection cachedCollection;
+        private readonly IMapperCompiler compiler;
+        public SimpleMapper() : this(new ExpressionTreeCompiler(), new CachedMapperCollection())
+        {
+        }
+
+        //for testing purposes
+        internal SimpleMapper(IMapperCompiler compiler) : this(compiler, new CachedMapperCollection())
```
Comment (RamanBut-Husaim) [inline Mapper/SimpleMapper.cs]: @MartinEmbers Please, do not create code that will be used only by tests. It is much better to redesign it to support testability.

File: Mapper.Tests/CacheTests/MapperUnitEqualityComparerTests/MapperUnitEqualityComparerTests.cs [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+            Assert.True(comparer.Equals(null,null));
+        }
+
+        [Test]
+        public void Equals_NullAndObjPassed_FalseReturned()
+        {
+            IEqualityComparer<IMappingUnit> comparer = CreateEqualityComparer();
+
+            Assert.False(comparer.Equals(null, CreateMappingUnit()));
```
Comment (RamanBut-Husaim) [inline Mapper.Tests/CacheTests/MapperUnitEqualityComparerTests/MapperUnitEqualityComparerTests.cs]: @MartinEmbers It is better to  introduce a variable in such cases `var actual = comparer.Equals(null, CreateMappingUnit());` and make the assertion.

**Comment by MartinEmbers** [comment]
Please, take a look

**Comment by RamanBut-Husaim** [comment]
@MartinEmbers Great work! Thank you!

**Pull request merged**